npe2/_plugin_manager.py

```diff
@@ -1,10 +1,28 @@
 from __future__ import annotations
 
+from npe2.manifest.io import LayerType, WriterContribution
+
 __all__ = ["plugin_manager", "PluginContext", "PluginManager"]  # noqa: F822
 import sys
+from collections import Counter
 from pathlib import Path
-from typing import TYPE_CHECKING, DefaultDict, Dict, Iterator, List, Set, Tuple, Union
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    DefaultDict,
+    Dict,
+    Iterator,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+from warnings import warn
+
+from intervaltree import IntervalTree
 
+from ._command_registry import execute_command
 from .manifest import PluginManifest
 
 if TYPE_CHECKING:
```

Comment:
you can throw this under `if TYPE_CHECKING` for full parametrized `IntervalTree` support:

```py
    from typing import Generic, TypeVar

    T = TypeVar("T")

    class Interval(tuple, Generic[T]):
        begin: int
        end: int
        data: T

    class IntrvlTree(IntervalTree[T]):
        def addi(self, begin: int, end: int, data: T) -> None:
            ...

        def __getitem__(self, index: Union[int, slice]) -> Set[Interval[T]]:
            ...
```

Reply:
Seems like I couldn't just throw this under TYPE_CHECKING, but maybe I'm fooling myself.

Reply:
Oh, I'd get a "not defined" constructor at runtime.  For example, running the tests gives

```
npe2/__init__.py:9: in <module>
    from ._plugin_manager import write_layers
npe2/_plugin_manager.py:73: in <module>
    class PluginManager:
npe2/_plugin_manager.py:81: in PluginManager
    ] = DefaultDict(TypedIntervalTree)
E   NameError: name 'TypedIntervalTree' is not defined
```

Reply:
oh also... the argument to `DefaultDict()` should _not_ be your typed version ... (just use it for the type hint).  So, something like this:

```py
from __future__ import annotations

...

_writers_by_type : DefaultDict[LayerType, TypedIntervalTree[WriterContribution]] = DefaultDict(IntervalTree)
```